diff --git a/nlw/lib/Socialtext/DisplayPlugin.pm b/nlw/lib/Socialtext/DisplayPlugin.pm
index 5eb4691..73d1bc3 100644
--- a/nlw/lib/Socialtext/DisplayPlugin.pm
+++ b/nlw/lib/Socialtext/DisplayPlugin.pm
@@ -14,8 +14,11 @@ use Socialtext::l10n qw/loc system_locale/;
 use Socialtext::Locales qw/available_locales/;
 use Socialtext::JSON;
 use Socialtext::Timer;
+use Socialtext::Paths;
+use Socialtext::File qw/get_contents_utf8 set_contents_utf8/;
 use Apache::Cookie;
 use Socialtext::Events;
+use File::Path qw/mkpath/;
 
 sub class_id { 'display' }
 const class_title => loc('Screen Layout');
@@ -272,9 +275,13 @@ sub _render_display {
         @{$self->hub->attachments->all(page_id => $page->id)},
     ];
 
+    my $frame_name = $self->_render_user_frame;
+
     return $self->template_render(
         template => 'view/page/display',
+        paths => [ $self->user_frame_path ],
         vars     => {
+            frame_name => $frame_name,
             $self->hub->helpers->global_template_vars,
             accept_encoding         => eval {
                 $self->hub->rest->request->header_in( 'Accept-Encoding' )
@@ -324,6 +331,46 @@ sub _render_display {
     );
 }
 
+sub user_frame_path {
+    return Socialtext::Paths::cache_directory('user_frame');
+}
+
+sub _render_user_frame {
+    my $self = shift;
+
+    my $frame_path = $self->user_frame_path;
+    my $user_id = $self->hub->current_user->user_id;
+    my $wksp_id = $self->hub->current_workspace->workspace_id;
+    $user_id =~ m/^(\d\d?)/;
+    my $user_prefix = $1;
+
+    my $frame_dir = "$frame_path/$user_prefix/$user_id";
+    my $tmpl_name = "/frame-ws-$wksp_id";
+    my $frame_tmpl = "$user_prefix/$user_id/$tmpl_name";
+    my $frame_file = "$frame_dir/$tmpl_name";
+
+    return $frame_tmpl if -f $frame_file;
+
+    warn "Rendering layout frame";
+    Socialtext::Timer->Continue('render_user_frame');
+    my $frame_content = $self->template_render(
+        template => 'layout/user_frame',
+        vars     => {
+            $self->hub->helpers->global_template_vars,
+            generate_user_frame => 1,
+        }
+    );
+
+    unless (-d $frame_dir) {
+        mkpath $frame_dir or die "Could not create $frame_dir: $!";
+    }
+
+    set_contents_utf8($frame_file, $frame_content);
+    warn "Wrote $frame_file";
+    Socialtext::Timer->Pause('render_user_frame');
+    return $frame_tmpl;
+}
+
 sub _get_minimal_page_info {
     my $self = shift;
     my $page = shift;
diff --git a/nlw/share/skin/s3/template/layout/user_frame b/nlw/share/skin/s3/template/layout/user_frame
new file mode 100644
index 0000000..ca918cd
--- /dev/null
+++ b/nlw/share/skin/s3/template/layout/user_frame
@@ -0,0 +1,191 @@
+[%# vim: set et sts=2 sw=2: -%]
+[%# This file will create a template that will later be rendered for a user %]
+[% TAGS <% %> %]
+<%- USE html_encode -%>
+<%- USE decorate -%>
+<%- USE label_ellipsis -%>
+[%- USE html_encode -%]
+[%- USE decorate -%]
+[%- USE label_ellipsis -%]
+<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd">
+<html xmlns="http://www.w3.org/1999/xhtml" xml:lang="en" lang="en">
+  <head>
+    <% IF firebug %>
+    <script type='text/javascript' src='http://getfirebug.com/releases/lite/1.2/firebug-lite-compressed.js'></script>
+    <% END %>
+
+    <% FILTER decorate "head" %>
+    <title>[% title || display_title %]<% IF workspace_present AND wiki.title %> / <% wiki.title %><% END %></title>
+    [% PROCESS element/css_links %]
+    [% IF feeds %]
+      [% IF feeds.rss %]
+        [% FOREACH feed IN feeds.rss %]
+    <link rel="alternate" type="application/rss+xml" title="[% feed.value.title | html_encode %]" href="[% feed.value.url %]" />
+        [% END %]
+      [% END %]
+
+      [% IF feeds.atom %]
+        [% FOREACH feed IN feeds.atom %]
+    <link rel="alternate" type="application/atom+xml" title="[% feed.value.title | html_encode %]" href="[% feed.value.url %]" />
+        [% END %]
+      [% END %]
+    <% END %>
+
+    <% IF loc_lang != 'en' %>
+    <script type="text/javascript" charset="utf-8"
+      src="<% wiki.skin_uri('s2') %>/javascript/l10ns.js.gz">
+    </script>
+    <% END %>
+
+    <script type="text/javascript" charset="utf-8"
+      src="<% wiki.skin_uri('s3') %>/javascript/socialtext-s3.js.gz">
+    </script>
+
+    [% IF head_template %]
+      [% INCLUDE $head_template %]
+    [% END %]
+
+    [% INCLUDE "element/wiki/js-bootstrap" %]
+
+    <% IF customjs %>
+    <script type="text/javascript" charset="utf-8" src="<% customjs %>"></script>
+    <% END %>
+
+    <meta name="keywords" content="<% loc('wiki, business wiki, document management, enterprise wiki, intranet wiki, knowledge management, online collaboration, open source wiki, portal wiki, hosted wiki, collaboration, social software, social networking, dashboard, project collaboration, team collaboration, wiki appliance, wiki software, community') %>"/>
+    [% END %]
+  </head>
+  <body id="[% id || 'socialtext-wiki' %]" onload="if (typeof Socialtext != 'undefined') Socialtext.body_loaded = true">
+    <% FILTER decorate('body_begin') %>
+    <% END %>
+    <div id="mainWrap">
+      <div id="header" class="st-wiki-nav">
+        <div id="logo"><a href="/"><% loc('Socialtext') %></a></div><!-- logo END -->
+        <div id="globalNav" class="st-wiki-nav-content">
+          <ul class="st-wiki-nav-actions">
+            <% UNLESS user.is_guest %>
+              <li>
+                <% loc('Logged in as [_1]', '<b>' _ user.username _ '</b>') | decorate('logged_in_as') %>
+              </li>
+            <% END %>
+            [% IF miki_url %]
+            <li id="st-wikinav-mobile"><a href="[% miki_url %]"><% loc('Mobile') %></a></li>
+            [% END %]
+            <li id="st-wikinav-setting"><a href="?action=users_settings"><% loc('Settings') %></a></li>
+            <li id="st-wikinav-help"><a href="/help/index.cgi" target="_blank"><% loc('Help') %></a></li>
+            <% IF user.is_guest %>
+              <% IF self_registration AND NOT checker.check_permission('self_join') %>
+                <li id="st-wikinav-login"><a href="/challenge?[% wiki.uri | html %]"><% loc('Log in') %></a></li>
+                <li class="last"><a href="/nlw/register.html?redirect_to=<% hub.current_workspace.uri %>"><% loc('Register') %></a></li>
+              <% ELSE %>
+                <li id="st-wikinav-login" class="last"><a href="/challenge?[% wiki.uri | html %]"><% loc('Log in') %></a></li>
+              <% END %>
+            <% ELSE %>
+              <li id="st-wikinav-logout" class="last"><a id="logout_btn" href="/nlw/submit/logout<% IF wiki.is_public %>?redirect_to=[% wiki.uri | html %]<% END %>"><% loc('Log out') %></a></li>
+            <% END %>
+          </ul>
+          <div>
+            <% INCLUDE element/search_form %>
+          </div>
+        </div><!-- globalNav END -->
+      </div><!-- header END -->
+      <div id="mainNav" class="st-wiki-subnav-dashboard">
+       <%- IF signals_only; PROCESS element/main_nav/signals; 
+           ELSE; PROCESS element/main_nav/standard; 
+           END
+       -%>
+      </div><!-- mainNav END -->
+
+      <div id="st-actions-bar">
+          [% IF wiki.system_status %]
+              <div id="st-system-status">[% wiki.system_status %]</div>
+          [% END %]
+      </div>
+
+[% IF content_id.match('^st-(homepage|editcontention|listview)$') %]
+      <div id="st-display-mode-container" class="st-content st-content-no-controls">  	
+        <div id="contentContainer">
+           [% content %]
+        </div>
+[% ELSE %]
+
+      <% IF workspace_present %> 
+      <!-- workspaceNav BEGIN -->
+      <div id="workspaceContainer">
+        <div id="workspaceNav">
+          <ul>
+            <li><% loc("Workspace") %>:
+              [% IF current_workspace.title != current_page.title || action %]
+              <a href="/<% current_workspace.name %>" id="workspaceName" class="here"><% current_workspace.title %></a>
+              [% ELSE %]
+              <span class="here"><% current_workspace.title %></span>
+              [% END %]
+              </li>
+            <% FILTER decorate('workspace_nav') %>
+            <li><a href="/<% current_workspace.name %>?action=recent_changes"><% loc("What's New") %></a></li>
+            <li><a href="/<% current_workspace.name %>?action=display_watchlist"><% loc("My Watchlist") %></a></li>
+            <li><a href="/<% current_workspace.name %>?action=weblog_display"><% loc("Weblogs") %></a></li>
+            <% IF is_workspace_admin %>
+            <li><a href="/<% current_workspace.name %>?action=users_invitation"><% loc("Add Members") %></a></li>
+            <% END %>
+            <% END %>
+          </ul>
+            <% IF checker.check_permission('edit') %>
+              <div id="newPageButton">
+                <a id="st-create-content-link" href="/<% current_workspace.name %>?action=new_page">+ <% loc("New Page") %></a>
+              </div>
+            <% ELSE %>
+              <div id="newPageButton" class="disabled">
+              &nbsp;
+              </div>
+            <% END %>
+        </div>
+      </div>
+      <!-- workspaceNav END -->
+      <% END %>
+      <div id="st-display-mode-container" class="st-content">  	
+        [% content %]
+[% END %]
+        <div id="footer">
+          <div id="footerLinks">
+            <ul>
+              <li class="first">
+                <a target="_blank" id="st-footer-link-socialtext" href="http://www.socialtext.com/">
+                  <% loc('Socialtext: Enterprise Social Software') %>
+                  v<% app_version %>
+                </a>
+              </li>
+              <li>
+                <a target="_blank" id="st-footer-link-community" href="http://www.socialtext.net/exchange/">
+                  <% loc('Community') %>
+                </a>
+              </li>
+            </ul>
+          </div><!-- footerLinks END -->
+          <div id="footerGlobal">
+            <ul>
+              [% IF miki_url %]
+                <li><a id="st-footer-link-mobile" href="[% miki_url %]"><% loc('Mobile') %></a></li>
+              [% END %]
+                <li><a id="st-footer-link-setting" href="?action=users_settings"><% loc('Settings') %></a></li>
+                <li><a id="st-footer-link-help" href="/help/index.cgi" target="_blank"><% loc('Help') %></a></li>
+              <% IF user.is_guest %>
+                <% IF self_registration %>
+                  <li><a id="st-footer-link-login" href="/challenge?[% wiki.uri | html %]"><% loc('Log in') %></a></li>
+                    <% UNLESS workspace_present AND checker.check_permission('self_join') %>
+                      <li class="last"><a id="st-footer-link-register" href="/nlw/register.html?redirect_to=<% hub.current_workspace.uri %>"><% loc('Register') %></a></li>
+                    <% END %>
+                <% ELSE %>
+                  <li  class="last"><a id="st-footer-link-login" href="/challenge?[% wiki.uri | html %]"><% loc('Log in') %></a></li>
+                <% END %>
+              <% ELSE %>
+                <li class="last"><a id="st-footer-link-logout" href="/nlw/submit/logout<% IF wiki.is_public %>?redirect_to=[% wiki.uri | html %]<% END %>"><% loc('Log out') %></a></li>
+              <% END %>
+            </ul>
+          </div><!-- footerGlobal END -->
+        </div><!-- footer END -->
+      </div><!-- contentWrap END -->
+    </div><!-- mainWrap END -->
+    <% FILTER decorate('body_end') %>
+    <% END %>
+  </body>
+</html>
diff --git a/nlw/share/skin/s3/template/view/page/display b/nlw/share/skin/s3/template/view/page/display
index dd9fbae..8594ae1 100644
--- a/nlw/share/skin/s3/template/view/page/display
+++ b/nlw/share/skin/s3/template/view/page/display
@@ -19,7 +19,7 @@
     [%- END -%]
 [%- END -%]
 
-[% WRAPPER layout/html id="contentPage" %]
+[% WRAPPER $frame_name id="contentPage" %]
 <div id="controls">
   <div id="st-editing-tools-edit">
     [% IF page_locked_for_user %]
